component/stream/stream.go

```diff
@@ -245,57 +176,74 @@ func (st *Stream) OpenFile(options internal.OpenFileOptions) (*handlemap.Handle,
-		fileKey := options.Name
-		if st.cacheOnFileHandle {
-			handle.ID = handlemap.HandleID(nextHandleID.Inc())
-			fileKey = strconv.FormatUint((uint64(handle.ID)), 10)
+		if st.cache.openHandles >= st.cache.handleLimit {
+			err = errors.New("handle limit exceeded")
+			log.Err("Stream::OpenFile : error %s [%s]", options.Name, err.Error())
+			return handle, err
 		}
-
-		flock := st.fileKeyLocks.Get(fileKey)
-		flock.Lock()
-		st.streamCache.addFileKey(fileKey)
-		block, exists, _ := st.fetchBlock(fileKey, handle, 0)
+		atomic.AddInt32(&st.cache.openHandles, 1)
+		cacheObj := handlemap.Cache{
+			DataBuffer: common.NewLRUCache(int64(st.cache.bufferSizePerHandle)),
+		}
+		handle.CacheObj = &cacheObj
+		block, exists, _ := st.getBlock(handle, 0)
 		// if it exists then we can just RUnlock since we didn't manipulate its data buffer
 		st.unlockBlock(block, exists)
-		st.streamCache.incrementHandles(fileKey)
-		flock.Unlock()
 	}
 	return handle, err
 }
 
-func (st *Stream) fetchBlock(fileKey string, handle *handlemap.Handle, offset int64) (*cacheBlock, bool, error) {
-	block, exists := st.streamCache.getBlock(fileKey, offset, handle.Size)
-	if !exists {
+func (st *Stream) getBlock(handle *handlemap.Handle, offset int64) (*common.CacheBlock, bool, error) {
+	blockSize := st.cache.blockSize
+	blockKeyObj := offset
+	handle.CacheObj.Lock()
+	block, found := handle.CacheObj.DataBuffer.Get(blockKeyObj)
+	if !found {
+		if (offset + blockSize) > handle.Size {
+			blockSize = handle.Size - offset
+		}
+		block = &common.CacheBlock{
+			StartIndex: offset,
+			EndIndex:   offset + blockSize,
+			Data:       make([]byte, blockSize),
+			Last:       (offset + blockSize) >= handle.Size,
+		}
+		block.Lock()
+		handle.CacheObj.DataBuffer.Put(blockKeyObj, block)
+		handle.CacheObj.Unlock()
 		// if the block does not exist fetch it from the next component
 		options := internal.ReadInBufferOptions{
 			Handle: handle,
-			Offset: block.startIndex,
-			Data:   block.data,
+			Offset: block.StartIndex,
+			Data:   block.Data,
 		}
 		_, err := st.NextComponent().ReadInBuffer(options)
 		if err != nil && err != io.EOF {
-			return nil, exists, err
+			return nil, false, err
 		}
+		return block, false, nil
+	} else {
+		handle.CacheObj.Unlock()
+		return block, true, nil
 	}
-	return block, exists, nil
 }
 
-func (st *Stream) copyCachedBlock(fileKey string, handle *handlemap.Handle, offset int64, data []byte) (int, error) {
+func (st *Stream) copyCachedBlock(handle *handlemap.Handle, offset int64, data []byte) (int, error) {
 	dataLeft := int64(len(data))
 	// counter to track how much we have copied into our request buffer thus far
 	dataRead := 0
 	// covers the case if we get a call that is bigger than the file size
 	for dataLeft > 0 && offset < handle.Size {
 		// round all offsets to the specific blocksize offsets
-		cachedBlockStartIndex := (offset - (offset % st.streamCache.blockSize))
+		cachedBlockStartIndex := (offset - (offset % st.cache.blockSize))
```

Comment:
this logic will not work for write streaming so when we break the code into smaller parts this getBlocka and copyCachedBlock shall be part of the implementation of that piece.

Reply:
sounds good - that will be in the write PR.